### ISSUE TITLE
npm5 Authentication Failure (E401) for 3rd Party Private Packages Registry

### ISSUE THREAD

**Issue by simonua** [npm/npm#16528]
#### I'm opening this issue because:

  - [ ] npm is crashing.
  - [ ] npm is producing an incorrect install.
  - [x] npm is doing something I don't understand.
  - [ ] Other (_see below for feature requests_):

#### What's going wrong?
We have successfully used Microsoft's Visual Studio Team Services Package Management for npm with all versions npm 4. When trying out 5.0.0-beta.36, we immediately get E401 error codes.

[npm debug log](https://gist.github.com/simonua/124625a26f5f48e9c2ad8de61d692f3a)

#### How can the CLI team reproduce the problem?
Unfortunately, due to the setup with a 3rd party registry, I cannot provide details for reproduction of the error.

### supporting information:

 - `npm -v` prints: 5.0.0-beta.36
 - `node -v` prints: v7.1.0
 - `npm config get registry` prints: https://connectionseducation.pkgs.visualstudio.com/_packaging/ConnectionsEducationNpm/npm/registry/
 - Windows, OS X/macOS, or Linux?: Windows 7
 - Network issues:
   - Geographic location where npm was run:
   - [ ] I use a proxy to connect to the npm registry.
   - [ ] I use a proxy to connect to the web.
   - [ ] I use a proxy when downloading Git repos.
   - [ ] I access the npm registry via a VPN
   - [ ] I don't use a proxy, but have limited or unreliable internet access.
 - Container:
   - [ ] I develop using Vagrant on Windows.
   - [ ] I develop using Vagrant on OS X or Linux.
   - [ ] I develop / deploy using Docker.
   - [ ] I deploy to a PaaS (Triton, Heroku).

Supplemental information for our .npmrc files:
```
registry=https://connectionseducation.pkgs.visualstudio.com/_packaging/ConnectionsEducationNpm/npm/registry/
always-auth=true
//connectionseducation.pkgs.visualstudio.com/_packaging/ConnectionsEducationNpm/npm/registry/:_authToken=<removed>
```

**Comment by simonua** [comment]
@zkat @zarenner, ping

**Comment by zarenner** [comment]
@simonua Thanks for bringing this to our attention. Did a quick investigation, looks like a couple issues in https://github.com/zkat/pacote

registry-key.js has "path: parsed.path" when I believe it should be "pathname: parsed.path", thus the path segment (which VSTS has) is getting dropped.

I think there's also a redirect auth issue - after fixing the above, from debug logs it looks like it's (incorrectly) forwarding credentials to the Azure Blob server hosting the tarball, which then responds with a message "Authentication information is not given in the correct format. Check the value of Authorization header."

I'll file issues in that project shortly, and hopefully someone (me or someone else) on our team can help contribute fixes unless @zkat gets to it first.

**Comment by zkat** [comment]
@zarenner I'll gladly take your patches! Thanks so much. You totally beat me to it. I literally went to lunch right after this and was gonna fix it when I got back, so I haven't even gotten started. You're awesome! Let me know if there's anything at all I can do, and I look forward to your PR 💃

**Comment by simonua** [comment]
Wow, that was fast! If beta 37 is live, I will give this a whirl tomorrow and report in this thread to close the loop. Thank you!

**Comment by zkat** [comment]
@simonua it's live 👍

**Comment by simonua** [comment]
Hi @zkat, I pulled beta39 and attempted another install. I got past the E401 but am now looking at an [E400 error](https://gist.github.com/simonua/618b57d5d2092476733b10bdc8ebc89c). I suspect this may be what @zarenner alluded to in the follow-up items, so this present issue here appears resolved since I no longer see an E401. =)

Many thanks again to both of you and your respective teams!

**Comment by zkat** [comment]
There's a community PR over at https://github.com/npm/node-fetch-npm/issues/1 now. This should be fixed once that lands.

**Comment by 4kochi** [comment]
Using `v5.0.0-beta.57` we still have the problem with our private registry. Here is the error log:

```
info it worked if it ends with ok
1 verbose cli [ '/usr/local/bin/node', '/usr/local/bin/npm5', 'i' ]
2 info using npm@5.0.0-beta.57
3 info using node@v7.10.0
4 notice BETA npm5 is beta software. Please see https://github.com/npm/npm/issues/16510 for known issues, and please file a new one if you find something new.
5 silly install loadCurrentTree
6 silly install readLocalPackageData
7 silly install loadIdealTree
8 silly install cloneCurrentTreeToIdealTree
9 silly install loadShrinkwrap
10 silly install loadAllDepsIntoIdealTree

.... 


1961 silly resolveWithNewModule rxjs@5.4.0 checking installable status
1962 silly pacote version manifest for sass-loader@6.0.5 fetched in 18ms
1963 silly resolveWithNewModule sass-loader@6.0.5 checking installable status
1964 silly pacote version manifest for source-map-loader@0.2.1 fetched in 23ms
1965 silly resolveWithNewModule source-map-loader@0.2.1 checking installable status
1966 silly pacote version manifest for strip-loader@0.1.2 fetched in 23ms
1967 silly resolveWithNewModule strip-loader@0.1.2 checking installable status
1968 silly pacote version manifest for ts-helpers@1.1.2 fetched in 23ms
1969 silly resolveWithNewModule ts-helpers@1.1.2 checking installable status
1970 silly pacote version manifest for tslib@1.7.1 fetched in 20ms
1971 silly resolveWithNewModule tslib@1.7.1 checking installable status
1972 silly pacote version manifest for tslint@5.2.0 fetched in 21ms
1973 silly resolveWithNewModule tslint@5.2.0 checking installable status
1974 silly pacote version manifest for tslint-loader@3.5.3 fetched in 22ms
1975 silly resolveWithNewModule tslint-loader@3.5.3 checking installable status
1976 silly pacote version manifest for typescript@2.3.2 fetched in 26ms
1977 silly resolveWithNewModule typescript@2.3.2 checking installable status
1978 http fetch GET 401 https://my.private.registry.com/artifactory/api/npm/npm/@types/node/-/node-7.0.21.tgz 78ms
1979 silly fetchPackageMetaData error for @types/node@7.0.21 401 Unauthorized: @types/node@7.0.21
1980 silly pacote range manifest for ulpian-azure-deploy@^0.2 fetched in 20ms
1981 silly resolveWithNewModule ulpian-azure-deploy@0.2.1557 checking installable status
1982 silly pacote version manifest for webpack@2.5.1 fetched in 21ms
1983 silly resolveWithNewModule webpack@2.5.1 checking installable status
1984 silly pacote version manifest for webpack-dev-server@2.4.5 fetched in 19ms
1985 silly resolveWithNewModule webpack-dev-server@2.4.5 checking installable status
1986 http fetch GET 401 https://my.private.registry.com/artifactory/api/npm/npm/zone.js/-/zone.js-0.8.11.tgz 53ms
1987 silly fetchPackageMetaData error for zone.js@0.8.11 401 Unauthorized: zone.js@0.8.11
1988 silly timing Tried to end timer that doesn't exist: stage:runTopLevelLifecycles
1989 verbose stack Error: 401 Unauthorized: @types/node@7.0.21
1989 verbose stack     at fetch.then.res (/usr/local/lib/node_modules/npm5/node_modules/pacote/lib/fetchers/registry/fetch.js:32:19)
1989 verbose stack     at tryCatcher (/usr/local/lib/node_modules/npm5/node_modules/bluebird/js/release/util.js:16:23)
1989 verbose stack     at Promise._settlePromiseFromHandler (/usr/local/lib/node_modules/npm5/node_modules/bluebird/js/release/promise.js:512:31)
1989 verbose stack     at Promise._settlePromise (/usr/local/lib/node_modules/npm5/node_modules/bluebird/js/release/promise.js:569:18)
1989 verbose stack     at Promise._settlePromise0 (/usr/local/lib/node_modules/npm5/node_modules/bluebird/js/release/promise.js:614:10)
1989 verbose stack     at Promise._settlePromises (/usr/local/lib/node_modules/npm5/node_modules/bluebird/js/release/promise.js:693:18)
1989 verbose stack     at Async._drainQueue (/usr/local/lib/node_modules/npm5/node_modules/bluebird/js/release/async.js:133:16)
1989 verbose stack     at Async._drainQueues (/usr/local/lib/node_modules/npm5/node_modules/bluebird/js/release/async.js:143:10)
1989 verbose stack     at Immediate.Async.drainQueues (/usr/local/lib/node_modules/npm5/node_modules/bluebird/js/release/async.js:17:14)
1989 verbose stack     at runCallback (timers.js:672:20)
1989 verbose stack     at tryOnImmediate (timers.js:645:5)
1989 verbose stack     at processImmediate [as _immediateCallback] (timers.js:617:5)
1990 verbose cwd /Users/pitti/web/ulpian-fallbearbeitung-ui
1991 verbose Darwin 16.6.0
1992 verbose argv "/usr/local/bin/node" "/usr/local/bin/npm5" "i"
1993 verbose node v7.10.0
1994 verbose npm  v5.0.0-beta.57
1995 error code E401
1996 error 401 Unauthorized: @types/node@7.0.21
1997 verbose exit [ 1, true ]
```

And here the local .npmrc in my project: 

```
registry=http://my.private.registry.com:8081/artifactory/api/npm/npm
email=user@domain.com
_auth=7647245gfhjewgfiwt4r8oqr7rw7qfho8qwfe
always-auth=true
```

**Comment by 4kochi** [comment]
Just an additional note. `npm 4.6.1` works fine.

**Comment by zkat** [comment]
@4kochi `_auth` was removed as of `npm@5.0.0`, so you'll need to log in to artifactory again, using `npm login --registry=<artifactory> --always-auth`. That should generate the necessary configuration to do `Basic` auth (to be clear: we still support basic auth, it's just not through `_auth` anymore).

**Comment by kjvalencik** [comment]
@zkat This isn't in the CHANGELOG. Also, `_auth` appears to be still supported for the majority of commands (e.g., `npm show`).

Was it intentionally removed or is it just an oversight? The best I can tell, this is the only place that doesn't support the v1 `_auth` field:

https://github.com/npm/npm/blob/latest/lib/config/pacote.js

**Comment by zkat** [comment]
ohmigosh I totally messed this up. We deprecated `_token` and I had a massive brain fart. Let's make sure npm5 supports `_auth`.

Sorry for the misunderstanding. I can take a look soon, or I'll take a patch if anyone is up for doing it. The change is fairly straightforward: add support for it in `lib/config/pacote.js` to make sure pacote gets it, then you have to go into https://github.com/zkat/pacote/blob/latest/lib/fetchers/registry/fetch.js#L86-L98 and add another case that checks for `auth.auth` and passes the base64 straight through as a header (instead of generating it itself, like what happens with `username` and `password`.

There's a second bug here, too: `lib/config/pacote.js` should include username, password, and auth if it's in the _global_ configuration with all `auth` objects it generates, unless that object already has its own values. That'll get the toplevel `username=...` stuff working, too.

If any of y'all are interested in picking this up, lmk. I think it's a relatively low-effort patch, but I'm also running around fixing everything _else_ so idk if I'll prioritize this (specially since there's workarounds).

Cheers~

**Comment by jozemlakar** [comment]
I'm on it. I'll let you handle the second bug unless I hit it when testing or someone explains better.

**Comment by jozemlakar** [comment]
Added PR #16879 to fix the reported issue.

**Comment by ottlinger** [comment]
Still an issue with `npm 5.0.1` - is there a release scheduled to fix the authentication issue?

**Comment by jozemlakar** [comment]
Yes. 5.0.3

**Comment by alexan** [comment]
I'm using 5.0.3 and _auth and username+_password login is still not working. 
4.6.1 is working fine

**Comment by jozemlakar** [comment]
yes, it was scheduled for 5.0.3. As you can read in #16879 there is a conflict with #17076

**Comment by uhrb** [comment]
This bug is amazing. Could you please consider private registries when you testing it?

**Comment by zkat** [comment]
@l0nley it was considered! Just people still auth into these in weird ways and I got confused between `_token` and `_auth` when removing the former :P

**Comment by simonua** [comment]
Hi @zkat, in addition to the unit tests you already have, does the cli team have test plans that could be peer-reviewed and amended? I know there are only so many folks working on your team, so that may be a tall order, but it could help shore things up and spread the load a little. That also means more community responsibility for FOSS. =P

**Comment by zkat** [comment]
@simonua we have no formal QA 'cause it's only two of us 😭

**Comment by mrchief** [comment]
@zkat v5.0.3 still doesn't support `_auth`. I tried `auth`, `_authToken` but no luck. What is the workaround? I cannot use npm login artifactory since our Artifactory is protected by Okta and using the apiKey as password doesn't work. What is the workaround that some of the comments seem to mention?

**Comment by jozemlakar** [comment]
For now: downgrade do npm@4

**Comment by mrchief** [comment]
> For now: downgrade do npm@4

Ironically, for any Artifactory or other private package management systems, `npm i npm@4.6.1 -g` also fails (if you upgraded to npm5 already). I had to temporarily modify `.npmrc` to point to public repo directly to bypass Artifcatory :)

**Comment by kjvalencik** [comment]
@mrchief You can work around the issue by using registry specific credentials instead of global ones.

```
registry=https://artifactory.example.com/artifactory/api/npm/your-npm-registry

//artifactory.example.com/artifactory/api/npm/:username=your-username
//artifactory.example.com/artifactory/api/npm/:password=your-password-base64-encoded
```

**Comment by mrchief** [comment]
@kjvalencik With your example, the request will still go thru Artifactory. And because npm@5 dropped `_auth`, Artifactory will not get the credentials so will fail with E401. Or am I missing something?

**Comment by kjvalencik** [comment]
@mrchief My example uses registry specific config instead of the global `_auth` config (which is currently broken). This example is *confirmed* working with artifactory.

**Comment by mrchief** [comment]
@kjvalencik Gotcha! I tried your way and now I get 403. I'm not sure if this is something caused by the way our Artifactory is setup or something I'm missing. But 401 is gone.

**Comment by itavy** [comment]
tested with Nexus 3.2.0, both scoped and unscoped packages

if you need to use it only as proxy:
```
registry=https://nexus.example.com/repository/group/
//https://nexus.example.com/repository/group/:username=<username>
//https://nexus.example.com/repository/group/:password=<base64 password encoded>
```
you must put the final / in all the lines

if you need to publish in the private repository you need to fill _auth field and all the other nexus requirements

**Comment by mrchief** [comment]
@kjvalencik Looks like our Artifactory (which is still on npm4) doesn't support it.

**Comment by wsmckenz** [comment]
So i'm having some of the same problem, but not with the _auth token itself. It is with always-auth, and even with the work around, it doesn't seem to work. I need to use auth on gets as well as puts. Publishing is working, but I can pull packages, because the auth header is not going into GET request.


.npmrc
```
prefix=c:\Users\wism\AppData\Roaming\npm-cache
email=William.Mckenzie@chevron.com
registry=https://cvxartifactorypro.azure.chevron.com/api/npm/npm/
auth-type=legacy
_auth=redacted
strict-ssl=false
always-auth=true
loglevel = silly
//https://cvxartifactorypro.azure.chevron.com/api/npm/npm/:always-auth=true
//https://cvxartifactorypro.azure.chevron.com/api/npm/npm/:username=wmckenzie
//https://cvxartifactorypro.azure.chevron.com/api/npm/npm/:password=redacated==
```

command: `npm install -g bower`

```
npm info it worked if it ends with ok
npm verb cli [ 'C:\\Program Files\\nodejs\\node.exe',
npm verb cli   'C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js',
npm verb cli   'prefix',
npm verb cli   '-g' ]
npm info using npm@5.0.3
npm info using node@v8.1.0
npm verb exit [ 0, true ]
npm info ok
npm info it worked if it ends with ok
npm verb cli [ 'C:\\Program Files\\nodejs\\node.exe',
npm verb cli   'C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js',
npm verb cli   'install',
npm verb cli   '-g',
npm verb cli   'bower' ]
npm info using npm@5.0.3
npm info using node@v8.1.0
npm verb npm-session 4c3aca39873f71bd
npm sill install loadCurrentTree
npm sill install readGlobalPackageData
npm http fetch GET 200 https://cvxartifactorypro.azure.chevron.com/api/npm/npm/bower 1146ms
npm http fetch GET 401 https://cvxartifactorypro.azure.chevron.com:443/api/npm/npm/bower/-/bower-1.8.0.tgz 54ms
npm sill fetchPackageMetaData error for bower@latest 401 Unauthorized: bower@latest
npm verb stack Error: 401 Unauthorized: bower@latest
npm verb stack     at fetch.then.res (C:\Program Files\nodejs\node_modules\npm\node_modules\pacote\lib\fetchers\registry\fetch.js:41:19)
npm verb stack     at tryCatcher (C:\Program Files\nodejs\node_modules\npm\node_modules\bluebird\js\release\util.js:16:23)
npm verb stack     at Promise._settlePromiseFromHandler (C:\Program Files\nodejs\node_modules\npm\node_modules\bluebird\js\release\promise.js:512:31)
npm verb stack     at Promise._settlePromise (C:\Program Files\nodejs\node_modules\npm\node_modules\bluebird\js\release\promise.js:569:18)
npm verb stack     at Promise._settlePromise0 (C:\Program Files\nodejs\node_modules\npm\node_modules\bluebird\js\release\promise.js:614:10)
npm verb stack     at Promise._settlePromises (C:\Program Files\nodejs\node_modules\npm\node_modules\bluebird\js\release\promise.js:693:18)
npm verb stack     at Async._drainQueue (C:\Program Files\nodejs\node_modules\npm\node_modules\bluebird\js\release\async.js:133:16)
npm verb stack     at Async._drainQueues (C:\Program Files\nodejs\node_modules\npm\node_modules\bluebird\js\release\async.js:143:10)
npm verb stack     at Immediate.Async.drainQueues (C:\Program Files\nodejs\node_modules\npm\node_modules\bluebird\js\release\async.js:17:14)
npm verb stack     at runCallback (timers.js:800:20)
npm verb stack     at tryOnImmediate (timers.js:762:5)
npm verb stack     at processImmediate [as _immediateCallback] (timers.js:733:5)
npm verb cwd D:\
npm verb Windows_NT 10.0.14393
npm verb argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "install" "-g" "bower"
npm verb node v8.1.0
npm verb npm  v5.0.3
npm ERR! code E401
npm ERR! 401 Unauthorized: bower@latest
npm verb exit [ 1, true ]

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\wism\AppData\Roaming\npm-cache\_logs\2017-06-21T18_08_53_796Z-debug.log
```

**Comment by simonua** [comment]
@wsmckenz, did you just publish your auth token?

**Comment by wsmckenz** [comment]
Yeah, my bad, had it in two places. No worries. It's not publicly accessible, i'll edit it.

**Comment by nileshmali** [comment]
Seems like npm team is not keen to fix this issue and is not part of next release.
This has broken our private repos and will be forced to switch to other alternatives.

**Comment by zkat** [comment]
Is this still broken in 5.0.4? I'm finally back from vacation so I'll be tackling this stuff more aggressively 💚

**Comment by jozemlakar** [comment]
@zkat yes. Needs #16879

**Comment by oshalygin** [comment]
Tracking this to see if what can be done.  Using Artifactory and things are failing for me.  Going to try @kjvalencik 's approach now

**Comment by oshalygin** [comment]
Running a few tests in my CI environment, downgrading via `npm install npm@4.0.6 -g` seems to get things back on track.  I just ssh'd into the box to test this so I'll need to do a full roundtrip to see if things are resolved.

I'll update this message with details.

EDIT: I ended up just using yarn instead to be honest.

**Comment by zkat** [comment]
This has been merged into `release-next` and will be going out with the next release. If you want to verify that the patches did what they're supposed to, you can install `npmc@5.0.4-canary.4` and try an install. Please let me know if that works for you, and we can close this issue. 👍 Cheers to @jozemlakar for the patch!

**Comment by ifuller1** [comment]
Works for me - thanks guys 👍 

What's the easiest way to track when this is in the next official npm release?

**Comment by phipsy7** [comment]
It also works for me. Thanks a lot.
We're using Nexus Repository Manager OSS version 3.0.0-03 for our private repos.

**Comment by rikkert** [comment]
I can confirm that everything works now as it did on npm4 with Artifactory.

**Comment by hswolff** [comment]
Can also confirm it works with Nexus Repository Manager OSS 2.14.4-03

**Comment by simonua** [comment]
@ifuller1, this is not quite what you're looking for, but you can compare _canary_ to _latest_ and get an indication of what's to come (thanks, Kat).

https://github.com/npm/npm/compare/latest...canary

**Comment by ackmanx** [comment]
Is there a release schedule or any idea when `release-next` comes out?

**Comment by zkat** [comment]
🤘 rock on. Glad this is fixed for y'all! Thanks for your patience. That vacation I took helped a lot ^_^

@ifuller1 @simonua You can use the canary to get the latest patches before we do an official npm release. The comparison you want to see what's actually gonna go into the official release is actually  https://github.com/npm/npm/compare/latest...release-next, since I will often put experimental code into the canary that isn't gonna be ready for release.

Anyway. Closing this now since it seems the general consensus is that yes, this fixed it. Thanks all for your patience and the detailed bug reports! Keep 'em coming~ (also the patches are nice. I like the patches a lot 💯 )

**Comment by oshalygin** [comment]
Thanks!!!! @zkat works for me as well

**Comment by casmith** [comment]
Confirmed working on Nexus 2.14.2. Thanks @zkat and @jozemlakar!

**Comment by roc** [comment]
@zkat `_auth`ome!

**Comment by f3l1x** [comment]
We are running Nexus 3.6.1 and NPM 5.5.1.

Adding `always-auth=true` to `.npmrc` helped!

**Comment by pavelstudeny** [comment]
npm 5.5.1 still doesn't work for me with artifactory 5.3.1. Shall I reopen this? Callstack is the same:
```
1597 verbose stack Error: 401 Unauthorized: @mycompany/my-module@^3.0.36
1597 verbose stack     at fetch.then.res (C:\Program Files\nodejs\node_modules\npm\node_modules\pacote\lib\fetchers\registry\fetch.js:42:19)
1597 verbose stack     at tryCatcher (C:\Program Files\nodejs\node_modules\npm\node_modules\bluebird\js\release\util.js:16:23)
1597 verbose stack     at Promise._settlePromiseFromHandler (C:\Program Files\nodejs\node_modules\npm\node_modules\bluebird\js\release\promise.js:512:31)
1597 verbose stack     at Promise._settlePromise (C:\Program Files\nodejs\node_modules\npm\node_modules\bluebird\js\release\promise.js:569:18)
1597 verbose stack     at Promise._settlePromise0 (C:\Program Files\nodejs\node_modules\npm\node_modules\bluebird\js\release\promise.js:614:10)
1597 verbose stack     at Promise._settlePromises (C:\Program Files\nodejs\node_modules\npm\node_modules\bluebird\js\release\promise.js:693:18)
1597 verbose stack     at Async._drainQueue (C:\Program Files\nodejs\node_modules\npm\node_modules\bluebird\js\release\async.js:133:16)
1597 verbose stack     at Async._drainQueues (C:\Program Files\nodejs\node_modules\npm\node_modules\bluebird\js\release\async.js:143:10)
1597 verbose stack     at Immediate.Async.drainQueues (C:\Program Files\nodejs\node_modules\npm\node_modules\bluebird\js\release\async.js:17:14)
1597 verbose stack     at runCallback (timers.js:789:20)
1597 verbose stack     at tryOnImmediate (timers.js:751:5)
1597 verbose stack     at processImmediate [as _immediateCallback] (timers.js:722:5)
```

**Comment by jozemlakar** [comment]
Callstack will always be the same if the server returns 401. The real question is WHY that is the case. And for that you need to see what the request held and what your config is. Post the npmrc.

It is working for most and us also, so there must be something more specific.

**Comment by webmutation** [comment]
5.8.0 also has this behavior from time to time on node 8.6.0 :/

**Comment by pavelstudeny** [comment]
forgot to post the .npmrc, apologies:
```
_auth=eE...a
init.author.name=My Name
init.author.email=my.email@domain.com
email=my.email@domain.com
@myscope:registry=https://internal-server/api/npm/npm/
@myscope:always-auth=false
```
(obfuscated the real values)

**Comment by rodrigoortegakds** [comment]
I has solved it making a log out explicitly to npm before. Maybe this procedure may help you:
```bash
npm logout
```

**Comment by webmutation** [comment]
Ok, I will give it a try and change our CI scripts to do a npm logout before doing a login... let's see

**Comment by k7sleeper** [comment]
Perhaps ```always-auth=true```?
